pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -290,45 +291,47 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListeners(env model.En
 
 	var tcpListeners, httpListeners []*xdsapi.Listener
 
-	wildcardListenerPorts := make(map[int]bool)
+	listenerMap := make(map[string]*xdsapi.Listener)
 	for _, service := range services {
 		for _, servicePort := range service.Ports {
 			clusterName := model.BuildSubsetKey(model.TrafficDirectionOutbound, "",
 				service.Hostname, servicePort)
 
+			listenAddress := WildcardAddress
 			var addresses []string
-			var listenAddress string
+			var listenerMapKey string
 			listenerOpts := buildListenerOpts{
 				env:            env,
 				proxy:          node,
 				proxyInstances: proxyInstances,
-				ip:             WildcardAddress,
 				port:           servicePort.Port,
 				protocol:       servicePort.Protocol,
 			}
-			var l *xdsapi.Listener
+
 			switch servicePort.Protocol {
 			case model.ProtocolTCP, model.ProtocolHTTPS, model.ProtocolMongo, model.ProtocolRedis:
-				if service.Resolution == model.Passthrough {
-					// ensure only one wildcard listener is created per port if its headless service
-					// or if this is in environment where services don't get a dummy load balancer IP.
-					if wildcardListenerPorts[servicePort.Port] {
-						log.Debugf("Multiple definitions for port %d", servicePort.Port)
-						continue
-					}
-					wildcardListenerPorts[servicePort.Port] = true
-					listenAddress = WildcardAddress
-					addresses = nil
-				} else {
+				if service.Resolution != model.Passthrough {
 					listenAddress = service.Address
 					addresses = []string{service.Address}
 				}
-				listenerOpts.ip = listenAddress
+
+				listenerMapKey = fmt.Sprintf("%s:%d", listenAddress, servicePort.Port)
+				if _, exists := listenerMap[listenerMapKey]; exists {
+					log.Warnf("Multiple TCP listener definitions for %s", listenerMapKey)
+					continue
+				}
+
 				listenerOpts.networkFilters = buildOutboundNetworkFilters(clusterName, addresses, servicePort)
-				l = buildListener(listenerOpts)
-				tcpListeners = append(tcpListeners, l)
-				// TODO: Set SNI for HTTPS
 			case model.ProtocolHTTP2, model.ProtocolHTTP, model.ProtocolGRPC:
+				listenerMapKey = fmt.Sprintf("%s:%d", listenAddress, servicePort.Port)
+				if l, exists := listenerMap[listenerMapKey]; exists {
+					if !strings.HasPrefix(l.Name, "http") {
```

Comment:
this is brittle. maybe we should store the protocol etc along with the listener.

Reply:
Yeah I didn’t like it either. Will try to rework in next iteration.